### PR TITLE
Added authentication

### LINKED ISSUE #19: Authentication service

### PR DESCRIPTION
**Pull request by vadzim-matsiuk**
close #19

### CODE REVIEW

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 1]
```diff
@@ -87,19 +92,35 @@ private Link createLink() {
...
         User user = new User();
         user.setUsername("author");
+        user.setEncryptedPassword(passwordEncoder.encode("password"));
         userRepository.save(user);
         return user;
     }
 
+    private HttpHeaders getHeadersWithToken() {
+        User user = new User();
+        user.setUsername("author");
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: extract variable as static field

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 2]
```diff
@@ -87,19 +92,35 @@ private Link createLink() {
...
+        user.setPassword("password");
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: extract variable as static field

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 3]
```diff
@@ -87,19 +92,35 @@ private Link createLink() {
...
     public void createWithGeneratedShortened() throws IOException {
         createUser();
+        HttpHeaders headers = getHeadersWithToken();
 
         LinkCreationDto link = mapper.readValue(new File("src/test/resources/LinkWithoutShortened.JSON"),
                 LinkCreationDto.class);
 
-        ResponseEntity<LinkGetterDto> responseEntity = restTemplate.postForEntity("/links",
-                link, LinkGetterDto.class);
+        HttpEntity <LinkCreationDto> entity = new HttpEntity<>(link, headers);
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: Use method without variable directly. Apply this to the whole file

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 4]
```diff
@@ -174,8 +212,15 @@ public void getLinksByTag() {
...
-        LinkGetterDto[] response = restTemplate.getForObject("/links/users/author",
-                LinkGetterDto[].class);
+        HttpHeaders headers = getHeadersWithToken();
+
+        HttpEntity entity = new HttpEntity(headers);
+
+        ResponseEntity<LinkGetterDto[]> responseEntity = restTemplate.exchange("/links/users/author",
+                HttpMethod.GET, entity, LinkGetterDto[].class);
+
+        assertThat(responseEntity.getBody(), is(notNullValue()));
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: Extract common code to private method

File: src/test/java/simplifier/controllers/UserControllerIT.java [hunk 5]
```diff
@@ -96,4 +95,35 @@ public void conflictIfUsernameNotUnique() {
...
+        ResponseEntity<String> responseEntity = restTemplate.postForEntity("/users/login",
+                user, String.class);
+
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.OK));
+        assertThat(responseEntity.getBody(), is(notNullValue()));
+    }
+
+    @Test
+    public void http422IfLoginInvalid() {
+        User user = new User();
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/UserControllerIT.java]: extract creating user to private method

File: src/test/java/simplifier/controllers/UserControllerIT.java [hunk 6]
```diff
@@ -96,4 +95,35 @@ public void conflictIfUsernameNotUnique() {
...
+        user.setUsername("test");
+        user.setPassword("test");
+
+        ResponseEntity<String> responseEntity = restTemplate.postForEntity("/users/login",
+                user, String.class);
+        String requiredMessage = "username or password is incorrect";
+        assertThat(responseEntity.getStatusCode(), is(HttpStatus.UNPROCESSABLE_ENTITY));
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/UserControllerIT.java]: extract common code to private method

File: src/main/java/simplifier/security/Http401UnauthorizedEntryPoint.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+public class Http401UnauthorizedEntryPoint implements AuthenticationEntryPoint {
+
+    @Override
+    public void commence(HttpServletRequest request,
+                         HttpServletResponse response,
+                         AuthenticationException authException) throws IOException, ServletException {
```
Comment (mikita-herasiutsin) [inline src/main/java/simplifier/security/Http401UnauthorizedEntryPoint.java]: ServletException is unnecessary

File: src/main/java/simplifier/security/TokenAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    private final String AUTH_HEADER = "Authorization";
+
+    private TokenHelper tokenHelper;
+
+    @Autowired
+    public void setTokenHelper(TokenHelper tokenHelper) {
+        this.tokenHelper = tokenHelper;
+    }
+
+    private String getToken(HttpServletRequest request) {
```
Comment (mikita-herasiutsin) [inline src/main/java/simplifier/security/TokenAuthenticationFilter.java]: private method should be below public ones

File: src/main/java/simplifier/security/TokenHelper.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Date;
+
+@Component
+public class TokenHelper {
+
+    @Value("${security.jwt.token.secret-key:secret-key}")
+    private String secretKey;
+
+    @Value("${security.jwt.token.expire-length:3600000}")
+    private long validityInMilliseconds = 3600000; // 1 hour
```
Comment (mikita-herasiutsin) [inline src/main/java/simplifier/security/TokenHelper.java]: refactor number to 3_600_000

File: src/main/java/simplifier/services/UserServiceImpl.java [hunk 10]
```diff
@@ -49,4 +66,12 @@ public void addLinkToUser(Link link) {
     public Optional<User> findByUsername(String username) {
         return userRepository.findByUsername(username);
     }
+
+    @Override
+    public String login(User user) {
+        String username = user.getUsername();
+        String password = user.getPassword();
+        authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(username, password));
```
Comment (mikita-herasiutsin) [inline src/main/java/simplifier/services/UserServiceImpl.java]: use getPassword() directly

File: src/test/java/simplifier/controllers/LinkControllerIT.java (hunk 3 above)
Comment (vadzim-matsiuk) [reply src/test/java/simplifier/controllers/LinkControllerIT.java]: I don't understand what you mean. Could you explain it to me?

Comment (mikita-herasiutsin) [reply src/test/java/simplifier/controllers/LinkControllerIT.java]: Use getHeaders() without creating variable. Creating variables is not necessary if it used once and method name is short

Comment (mikita-herasiutsin) [reply src/test/java/simplifier/controllers/LinkControllerIT.java]: Use getHeadersWithToken() without creating variable. Creating variables is not necessary if it used once and method name is too long

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 11]
```diff
@@ -164,26 +196,32 @@ public void conflictIfShortenedNotUnique() throws Exception {
     @Test
     public void getLinksByTag() {
         Link link = createLink();
-        LinkGetterDto[] response = restTemplate.getForObject("/links/tags/1",
-                LinkGetterDto[].class);
-        assertThat(response.length, is(1));
-        LinkGetterDto responseLink = response[0];
+
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: extract the whole code of this and getLinksByUser() methods to common one with 1 parameter (rest url)

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 12]
```diff
@@ -117,8 +141,10 @@ public void createWithExistShortened() throws IOException {
         LinkCreationDto link = mapper.readValue(new File("src/test/resources/validLink.JSON"),
                 LinkCreationDto.class);
 
-        ResponseEntity<LinkGetterDto> responseEntity = restTemplate.postForEntity("/links",
-                link, LinkGetterDto.class);
+        HttpEntity <LinkCreationDto> entity = new HttpEntity<>(link, getHeaders());
+
+        ResponseEntity<LinkGetterDto> responseEntity = restTemplate.exchange("/links", HttpMethod.POST,
+                entity, LinkGetterDto.class);
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: variable requiredMessage is not necessary in this method. Apply this to all code

File: src/test/java/simplifier/controllers/LinkControllerIT.java [hunk 13]
```diff
@@ -48,6 +46,12 @@
 
     private ObjectMapper mapper = new ObjectMapper();
 
+    private BCryptPasswordEncoder passwordEncoder;
+
+    private static final String USERNAME = "author";
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: Extract all variables used more that once to static fields

Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/LinkControllerIT.java]: Static fields should go first in classes. More information [here](http://www.oracle.com/technetwork/java/codeconventions-150003.pdf)

File: src/test/java/simplifier/controllers/UserControllerIT.java [hunk 14]
```diff
@@ -96,4 +96,28 @@ public void conflictIfUsernameNotUnique() {
         assertThat(responseEntity.getStatusCode(), is(HttpStatus.CONFLICT));
         assertThat(responseEntity.getBody(), is(requiredMessage));
     }
+
+    @Test
+    public void login() {
```
Comment (mikita-herasiutsin) [inline src/test/java/simplifier/controllers/UserControllerIT.java]: Rename to more informative shouldLoginWithOkStatusAndReturnToken. The name of a test method should  describe expected input and the expected result for that input. Apply this naming to all code

**Pull request merged**